File: study/lua/src/app/scene/view/campRace/CampRaceRankNode.lua
--排行榜
local CampRaceRankNode = class("CampRaceRankNode")
local CampRaceRankCell = require("app.scene.view.campRace.CampRaceRankCell")
local CampRaceHelper = require("app.scene.view.campRace.CampRaceHelper")

function CampRaceRankNode:ctor(target)
    self._target = target
    self._imageCountry = nil
    self._imageRankTitle = nil
    self._listRank = nil
    self._textMyPoint = nil
    self._textMyRank = nil
    self._textMyName = nil
    self:_init()
end

function CampRaceRankNode:_init()
    self._imageCountry = ccui.Helper:seekNodeByName(self._target, "ImageCountry")
    self._imageRankTitle = ccui.Helper:seekNodeByName(self._target, "ImageRankTitle")
    self._listRank = ccui.Helper:seekNodeByName(self._target, "ListRank")
    cc.bind(self._listRank, "ScrollView")
    self._listRank:setTemplate(CampRaceRankCell)
    self._listRank:setCallback(handler(self, self._onItemUpdate), handler(self, self._onItemSelected))
    self._listRank:setCustomCallback(handler(self, self._onItemTouch))
    self._textMyPoint = ccui.Helper:seekNodeByName(self._target, "TextMyPoint")
    self._textMyRank = ccui.Helper:seekNodeByName(self._target, "TextMyRank")
    self._textMyName = ccui.Helper:seekNodeByName(self._target, "TextMyName")

    self:_setMyInfo()
    self:_setTitle()
end

function CampRaceRankNode:onEnter()
end

function CampRaceRankNode:onExit()
end

function CampRaceRankNode:_setMyInfo()
    local myName = G_UserData:getBase():getName()
    self._textMyName:setString(myName)
    local myOfficerLevel = G_UserData:getBase():getOfficer_level()
    self._textMyName:setColor(Colors.getOfficialColor(myOfficerLevel)) 
end

function CampRaceRankNode:_setTitle()
    local camp = G_UserData:getCampRaceData():getMyCamp()

    local smallCamps = {4, 1, 3, 2}
	local campSmall = Path.getTextSignet("img_com_camp0"..smallCamps[camp])
    self._imageCountry:loadTexture(campSmall)
    
    local titleEnd = {"", "b", "c", "d"}
    local titleText = Path.getTextCampRace("txt_camp_01"..titleEnd[camp])
    self._imageRankTitle:loadTexture(titleText)
end

function CampRaceRankNode:setRankData(rankData)
    self._rankData = rankData
    self._listRank:clearAll()
    self._listRank:resize(#rankData)
end

function CampRaceRankNode:refreshMyRank()
    local camp = G_UserData:getCampRaceData():getMyCamp()
    local preRankData = G_UserData:getCampRaceData():getPreRankWithCamp(camp)
    local myRank = preRankData:getSelf_rank()
    self._textMyRank:setString(myRank)
    local myPoint = preRankData:getSelf_score()
    self._textMyPoint:setString(myPoint)
end

function CampRaceRankNode:_onItemUpdate(item, index)
    item:updateUI(index+1, self._rankData[ index+1 ] )
end

function CampRaceRankNode:_onItemSelected(item, index)
end

return CampRaceRankNode

File: study/ts/script/scene/view/campRace/CampRaceRankNode.ts
import { Colors, G_UserData } from "../../../init";
import CommonListView from "../../../ui/component/CommonListView";
import { handler } from "../../../utils/handler";
import { Path } from "../../../utils/Path";
import UIHelper from "../../../utils/UIHelper";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CampRaceRankNode extends cc.Component {
    @property({ type: cc.Label, visible: true })
    _textMyPoint: cc.Label = null;
    @property({ type: cc.Label, visible: true })
    _textMyRank: cc.Label = null;
    @property({ type: cc.Label, visible: true })
    _textMyName: cc.Label = null;
    @property({ type: cc.Sprite, visible: true })
    _imageCountry: cc.Sprite = null;
    @property({ type: cc.Sprite, visible: true })
    _imageRankTitle: cc.Sprite = null;
    @property({ type: CommonListView, visible: true })
    _listRank: CommonListView = null;
    @property(cc.Prefab)
    campRaceRankCell: cc.Prefab = null;
    _rankData: any;

    init() {
        this._listRank.init(this.campRaceRankCell, handler(this, this._onItemUpdate))
        this._setMyInfo();
        this._setTitle();
    }
    _setMyInfo() {
        var myName = G_UserData.getBase().getName();
        this._textMyName.string = (myName);
        var myOfficerLevel = G_UserData.getBase().getOfficer_level();
        this._textMyName.node.color = (Colors.getOfficialColor(myOfficerLevel));
    }
    _setTitle() {
        var camp = G_UserData.getCampRaceData().getMyCamp();
        var smallCamps = [
            4,
            1,
            3,
            2
        ];
        var campSmall = Path.getTextSignet('img_com_camp0' + smallCamps[camp]);
        UIHelper.loadTexture(this._imageCountry, campSmall);
        var titleEnd = [
            '',
            'b',
            'c',
            'd'
        ];
        var titleText = Path.getTextCampRace('txt_camp_01' + titleEnd[camp]);
        UIHelper.loadTexture(this._imageRankTitle, titleText);
    }
    setRankData(rankData) {
        this._rankData = rankData;
        this._listRank.setData(rankData.length);
    }
    refreshMyRank() {
        var camp = G_UserData.getCampRaceData().getMyCamp();
        var preRankData = G_UserData.getCampRaceData().getPreRankWithCamp(camp);
        var myRank = preRankData.getSelf_rank();
        this._textMyRank.string = (myRank);
        var myPoint = preRankData.getSelf_score();
        this._textMyPoint.string = (myPoint);
    }
    _onItemUpdate(item, index) {
        item.updateItem(index, this._rankData[index]);
    }
    _onItemSelected(item, index) {
    }
}